hugr-core/src/import.rs

```diff
@@ -0,0 +1,1214 @@
+        }
+
+        Ok(())
+    }
+
+    fn import_adt_and_rest(
+        &mut self,
+        node_id: model::NodeId,
+        ports: &'a [model::Port<'a>],
+    ) -> Result<(Vec<TypeRow>, TypeRow), ImportError> {
+        let Some((first, rest)) = ports.split_first() else {
+            return Err(model::ModelError::InvalidRegions(node_id).into());
+        };
+
+        let sum_rows: Vec<_> = {
+            let Some(term) = first.r#type else {
+                return Err(error_uninferred!("port type"));
+            };
+
+            let model::Term::Adt { variants } = self.get_term(term)? else {
+                return Err(model::ModelError::TypeError(term).into());
+            };
+
+            self.import_type_rows(*variants)?
+        };
+
+        let rest = self.get_port_types(rest)?;
+
+        Ok((sum_rows, rest))
+    }
+
+    fn import_tail_loop(
+        &mut self,
+        node_id: model::NodeId,
+        parent: Node,
+    ) -> Result<Node, ImportError> {
+        let node_data = self.get_node(node_id)?;
+        assert!(matches!(node_data.operation, model::Operation::TailLoop));
+
+        let [region] = node_data.regions else {
+            return Err(model::ModelError::InvalidRegions(node_id).into());
+        };
+        let region_data = self.get_region(*region)?;
+
+        let (sum_rows, rest) = self.import_adt_and_rest(node_id, region_data.targets)?;
+
+        let (just_inputs, just_outputs) = {
+            let mut sum_rows = sum_rows.into_iter();
+
+            let term = region_data.targets[0].r#type.unwrap();
+
+            let Some(just_inputs) = sum_rows.next() else {
+                return Err(model::ModelError::TypeError(term).into());
+            };
+
+            let Some(just_outputs) = sum_rows.next() else {
+                return Err(model::ModelError::TypeError(term).into());
+            };
+
+            (just_inputs, just_outputs)
+        };
+
+        let optype = OpType::TailLoop(TailLoop {
+            just_inputs,
+            just_outputs,
+            rest,
+            extension_delta: ExtensionSet::new(),
+        });
+
+        let node = self.make_node(node_id, optype, parent)?;
+
+        self.import_dfg_region(node_id, *region, node)?;
+        Ok(node)
+    }
+
+    fn import_conditional(
+        &mut self,
+        node_id: model::NodeId,
+        parent: Node,
+    ) -> Result<Node, ImportError> {
+        let node_data = self.get_node(node_id)?;
+        assert!(matches!(node_data.operation, model::Operation::Conditional));
+
+        let (sum_rows, other_inputs) = self.import_adt_and_rest(node_id, node_data.inputs)?;
+        let outputs = self.get_port_types(node_data.outputs)?;
+
+        let optype = OpType::Conditional(Conditional {
+            sum_rows,
+            other_inputs,
+            outputs,
+            extension_delta: ExtensionSet::new(),
+        });
+
+        let node = self.make_node(node_id, optype, parent)?;
+
+        for region in node_data.regions {
```

Comment:
In the model, everything order sensitive has been addressed by regions:
 - Input and Output are replaced by the data flow region's sources and targets
 - The first block and the Exit block are replaced by the control flow region's sources and targets
 - Instead of case nodes, the conditional node has multiple regions, and regions are ordered.

For now, the import and export convert to the order sensitive encoding of the core which is unchanged. How to proceed with that relates to https://github.com/CQCL/hugr/issues/1546 and is probably out of scope of this PR.